### PR TITLE
BUG: Allow non-callable attributes in aggregate function. Fixes GH16405

### LINKED ISSUE #16405: BUG: DataFrame.agg({'col': 'size'}) not working
`size` aggregation does not seem to work properly when used with ungrouped dataframes. When replacing `size` with `count` the examples run through. However, this change null semantics, counting nulls (`size`) in contrast to skipping nulls (`count`). The examples below are for Pandas version `0.20.1`.

When used with `.groupby(...)`, size aggregation works as expected:

```python
>>> pd.DataFrame({'g': [0, 0, 1], 'v': [1, 2, None]}).groupby('g').agg({'v': 'size'})
   v
g   
0  2
1  1
```

Without `.groupby`, it raises an recursion error:

```python
>>> pd.DataFrame({'g': [0, 0, 1], 'v': [1, 2, None]}).agg({'v': 'size'})
---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
<ipython-input-15-5fc5c38cdb5f> in <module>()
----> 1 pd.DataFrame({'g': [0, 0, 1], 'v': [1, 2, 3]}).agg({'v': 'size'})

/Volumes/Home/venvs/py3-data-science/lib/python3.5/site-packages/pandas/core/frame.py in aggregate(self, func, axis, *args, **kwargs)
   4250                 pass
   4251         if result is None:
-> 4252             return self.apply(func, axis=axis, args=args, **kwargs)
   4253         return result
   4254 

/Volumes/Home/venvs/py3-data-science/lib/python3.5/site-packages/pandas/core/frame.py in apply(self, func, axis, broadcast, raw, reduce, args, **kwds)
   4322         # dispatch to agg
   4323         if axis == 0 and isinstance(func, (list, dict)):
-> 4324             return self.aggregate(func, axis=axis, *args, **kwds)
   4325 
   4326         if len(self.columns) == 0 and len(self.index) == 0:

... last 2 frames repeated, from the frame below ...

/Volumes/Home/venvs/py3-data-science/lib/python3.5/site-packages/pandas/core/frame.py in aggregate(self, func, axis, *args, **kwargs)
   4250                 pass
   4251         if result is None:
-> 4252             return self.apply(func, axis=axis, args=args, **kwargs)
   4253         return result
   4254 

RecursionError: maximum recursion depth exceeded
```

### PR DESCRIPTION
**Pull request by pvomelveny**
- [x] closes #16405
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Can you add the test from the original issue here? It can go in `pandas/tests/frame/test_apply.py`

Also could use a release note in `doc/source/whatsnew/v0.20.2.txt`, under the bugfix section.

**Comment by pvomelveny** [comment]
I may have gone a bit overboard with the tests...

File: pandas/core/base.py [hunk 1]
```diff
@@ -378,7 +378,7 @@ def aggregate(self, func, *args, **kwargs):
     def _try_aggregate_string_function(self, arg, *args, **kwargs):
         """
         if arg is a string, then try to operate on it:
-        - try to find a function on ourselves
+        - try to find an attribute on ourselves
```
Comment (jreback) [inline pandas/core/base.py]: function (or attribute)

File: doc/source/whatsnew/v0.20.2.txt [hunk 2]
```diff
@@ -83,6 +83,7 @@ Reshaping
 - Bug in ``DataFrame.stack`` with unsorted levels in MultiIndex columns (:issue:`16323`)
 - Bug in ``pd.wide_to_long()`` where no error was raised when ``i`` was not a unique identifier (:issue:`16382`)
 - Bug in ``Series.isin(..)`` with a list of tuples (:issue:`16394`)
+- Bug in ``DataFrame.agg`` and ``Series.agg`` with aggregating on non-callable attributes (:issue:`16404`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.2.txt]: I *think* this only affects ``.size``, so say that

File: pandas/core/base.py [hunk 3]
```diff
@@ -387,7 +387,12 @@ def _try_aggregate_string_function(self, arg, *args, **kwargs):
 
         f = getattr(self, arg, None)
         if f is not None:
-            return f(*args, **kwargs)
+            if callable(f):
+                return f(*args, **kwargs)
+            # people may try to aggregate on a non-callable attribute
```
Comment (jreback) [inline pandas/core/base.py]: blank line

File: pandas/tests/frame/test_apply.py [hunk 4]
```diff
@@ -635,3 +635,46 @@ def test_nuiscance_columns(self):
...
+        # Mix function and non-function aggs
+        result1 = df.agg(['count', 'size'])
+        result2 = df.agg({'A': ['count', 'size'],
+                          'B': ['count', 'size'],
+                          'C': ['count', 'size']})
+        expected = pd.DataFrame({'A': {'count': 2, 'size': 3},
+                                 'B': {'count': 2, 'size': 3},
+                                 'C': {'count': 2, 'size': 3}})
+
+        assert_frame_equal(result1.reindex_like(expected),
```
Comment (jreback) [inline pandas/tests/frame/test_apply.py]: you can pass ``check_like=True`` to do this reindex comparisons.

File: pandas/tests/frame/test_apply.py [hunk 5]
```diff
@@ -635,3 +635,46 @@ def test_nuiscance_columns(self):
...
+                           result2.reindex_like(expected))
+        assert_frame_equal(result2.reindex_like(expected), expected)
+
+        # Just functional string arg is same as calling df.arg()
```
Comment (jreback) [inline pandas/tests/frame/test_apply.py]: can you add a minimal tests to ``series/test_apply.py``

File: pandas/tests/frame/test_apply.py [hunk 6]
```diff
@@ -635,3 +635,46 @@ def test_nuiscance_columns(self):
...
+        result = df.agg('count')
+        expected = df.count()
+
+        assert_series_equal(result, expected)
+
+        # Trying to to the same w/ non-function tries to pass args
+        # which we explicitly forbid
```
Comment (jreback) [inline pandas/tests/frame/test_apply.py]: this actually should work, what are the args/kwargs that are actually there? (ithere *maybe* internal ones that need to be remove FYI, e.g. ``_level`` or something)

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.20.2.txt (hunk 2 above)
Comment (pvomelveny) [inline doc/source/whatsnew/v0.20.2.txt]: As it works right now, you really can use *any* attribute in the aggregation, so something like:
```python
In [20]: df
Out[20]:
     A    B     C
0  NaN  1.0   foo
1  2.0  NaN  None
2  3.0  3.0   bar

In [21]: df.agg({'A': ['count', 'size', 'name'],
    ...:         'B': ['count', 'size', 'name', 'data'],
    ...:         'C': 'size'})
    ...:
Out[21]:
         A                B    C
count    2                2  NaN
data   NaN  [1.0, nan, 3.0]  NaN
name     A                B  NaN
size     3                3  3.0
```
works. 

If it should only affect size, we could change the check in base.py to just see if the arg passed is `'size'`.

Comment (pvomelveny) [inline doc/source/whatsnew/v0.20.2.txt]: This one's not actually outdated. I just messed up the bug # and had to fix it. The question of this only affecting size is still open

**Comment by TomAugspurger** [comment]
@jreback are you OK with this? I believe the discussion was whether we could have a simpler implementation by special-casing `'size'`?

**Comment by jreback** [comment]
can't change size as it's already defined as a property (for numpy compat)

File: doc/source/whatsnew/v0.20.2.txt [hunk 7]
```diff
@@ -89,6 +89,7 @@ Reshaping
 - Bug in ``pd.wide_to_long()`` where no error was raised when ``i`` was not a unique identifier (:issue:`16382`)
 - Bug in ``Series.isin(..)`` with a list of tuples (:issue:`16394`)
 - Bug in construction of a ``DataFrame`` with mixed dtypes including an all-NaT column. (:issue:`16395`)
+- Bug in ``DataFrame.agg`` and ``Series.agg`` with aggregating on non-callable attributes (:issue:`16405`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.2.txt]: ``DataFrame.agg()`` and ``Series.agg()``

File: pandas/tests/frame/test_apply.py [hunk 8]
```diff
@@ -635,3 +635,46 @@ def test_nuiscance_columns(self):
         expected = DataFrame([[6, 6., 'foobarbaz']],
                              index=['sum'], columns=['A', 'B', 'C'])
         assert_frame_equal(result, expected)
+
+    def test_non_callable_aggregates(self):
+
```
Comment (jreback) [inline pandas/tests/frame/test_apply.py]: can you add a comment of ``.size`` here (and how its a property of Series/DataFrame and hence we allow this)

**Review by jreback — APPROVED** [review]
very small doc change. ping on green.

**Comment by jreback** [comment]
thanks!

**Pull request merged**